### ISSUE TITLE
The RDB  file is damaged, how should I recover it?

### ISSUE THREAD

**Issue by zhangliang-zl** [RedisGraph/RedisGraph#2811]
RDB recovery fails after redisgraph crashes, how should I recover data?

Under what circumstances would this problem arise? How should I avoid it?

redis.log 

``` 
*** Preparing to test memory region 7fcbdbcda000 (20480 bytes)
*** Preparing to test memory region 7fcbdbce4000 (8192 bytes)
*** Preparing to test memory region 7fcbdbce6000 (4096 bytes)
*** Preparing to test memory region 7fcbdbce9000 (4096 bytes)
.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O
Fast memory test PASSED, however your memory can still be broken. Please run a memory test for several hours if possible.

------ DUMPING CODE AROUND EIP ------
Symbol: RdbLoadEdges_v9 (base: 0x7fcbd2de3ba0)
Module: /data/s/services/redisgraph/16479/server/so/redisgraph.so (base 0x7fcbd2bf4000)
$ xxd -r -p /tmp/dump.hex /tmp/dump.bin
$ objdump --adjust-vma=0x7fcbd2de3ba0 -D -b binary -m i386:x86-64 /tmp/dump.bin
------
25812:M 30 Dec 2022 17:07:20.436 # dump of function (hexdump of 246 bytes):
415741564155415455534883ec7848895424184885d20f84f10000004c8b25b5980a014889fb4989f64531ed4c8d7c24200f1f80000000004889df41ff14244889df4889c541ff14244889df488944240841ff14244889df4889042441ff142431f6498b3e4889ea4989c1498b8688000000488b40184a833cc800400f95c64883ec0841574c8b442410488b4c2418e8fce1ffff4889df41ff14245a594885c0745c8944241431ed0f1f8400000000004889df83c50141ff14244889df48890424e81afcffff0fb734244c89ff4889d14989c14889c24c894c240848890c24e80cbeffff488b0c244c8b4c24084889ce4c89cfe8c85f

=== REDIS BUG REPORT END. Make sure to include from START to END. ===

       Please report the crash by opening an issue on github:

           http://github.com/redis/redis/issues

  Suspect RAM error? Use redis-server --test-memory to verify it.

2206:C 30 Dec 2022 17:07:38.680 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
2206:C 30 Dec 2022 17:07:38.681 # Redis version=6.0.16, bits=64, commit=00000000, modified=0, pid=2206, just started
2206:C 30 Dec 2022 17:07:38.681 # Configuration loaded
2206:M 30 Dec 2022 17:07:38.685 * Running mode=standalone, port=16479.
2206:M 30 Dec 2022 17:07:38.685 # Server initialized
2206:M 30 Dec 2022 17:07:38.690 * <graph> Starting up RedisGraph version 2.8.6.
2206:M 30 Dec 2022 17:07:38.692 * <graph> Thread pool created, using 24 threads.
2206:M 30 Dec 2022 17:07:38.692 * <graph> Maximum number of OpenMP threads set to 24
2206:M 30 Dec 2022 17:07:38.692 * Module 'graph' loaded from /data/s/services/redisgraph/16479/server/so/redisgraph.so
2206:M 30 Dec 2022 17:07:38.693 * Loading RDB produced by version 6.0.16
2206:M 30 Dec 2022 17:07:38.693 * RDB age 7175 seconds
2206:M 30 Dec 2022 17:07:38.693 * RDB memory usage when created 4022.57 Mb


=== REDIS BUG REPORT START: Cut & paste starting from here ===
2206:M 30 Dec 2022 17:09:12.859 # Redis 6.0.16 crashed by signal: 11, si_code: 2
2206:M 30 Dec 2022 17:09:12.859 # Crashed running the instruction at: 0x7fc0af9e3c16
2206:M 30 Dec 2022 17:09:12.859 # Accessing address: 0x7fc0b7f46914
2206:M 30 Dec 2022 17:09:12.859 # Failed assertion: <no assertion failed> (<no file>:0)

------ STACK TRACE ------
EIP:
/data/s/services/redisgraph/16479/server/so/redisgraph.so(RdbLoadEdges_v9+0x76)[0x7fc0af9e3c16]

Backtrace:
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479(logStackTrace+0x2e)[0x47a5ee]
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479(sigsegvHandler+0xa9)[0x47acb9]
/lib64/libpthread.so.0(+0xf5d0)[0x7fc0b7d4e5d0]
/data/s/services/redisgraph/16479/server/so/redisgraph.so(RdbLoadEdges_v9+0x76)[0x7fc0af9e3c16]
/data/s/services/redisgraph/16479/server/so/redisgraph.so(RdbLoadGraph_v9+0x2f3)[0x7fc0af9e4013]
/data/s/services/redisgraph/16479/server/so/redisgraph.so(+0x1edca5)[0x7fc0af9e1ca5]
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479[0x45b350]
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479(rdbLoadRio+0x248)[0x45d0e8]
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479(rdbLoad+0x4d)[0x45daad]
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479(loadDataFromDisk+0x8c)[0x43b5cc]
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479(main+0x470)[0x42d020]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7fc0b79943d5]
/data/s/services/redisgraph/16479/server/bin/redis-server *:16479[0x42d359]

------ INFO OUTPUT ------
# Server
redis_version:6.0.16
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:567eea098253f33e
redis_mode:standalone
os:Linux 3.10.0-957.el7.x86_64 x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:atomic-builtin
gcc_version:9.3.1
process_id:2206
run_id:c5b6e2cec72b01dc75ea60f19f85dbcb569356f0
tcp_port:16479
uptime_in_seconds:94
uptime_in_days:0
hz:50
configured_hz:50
lru_clock:11446874
executable:/data/s/services/redisgraph/16479/server/bin/redis-server
config_file:/data/s/services/redisgraph/16479/data/redis.conf
io_threads_active:0

# Clients
connected_clients:2
client_recent_max_input_buffer:0
client_recent_max_output_buffer:0
blocked_clients:0
tracking_clients:0
clients_in_timeout_table:0

# Memory
used_memory:3981615456
used_memory_human:3.71G
used_memory_rss:0
used_memory_rss_human:0B
used_memory_peak:3981615456
used_memory_peak_human:3.71G
used_memory_peak_perc:100.07%
used_memory_overhead:839176
used_memory_startup:837848
used_memory_dataset:3980776280
used_memory_dataset_perc:100.00%
allocator_allocated:0
allocator_active:0
allocator_resident:0
total_system_memory:135025930240
total_system_memory_human:125.75G
used_memory_lua:37888
used_memory_lua_human:37.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:8589934592
maxmemory_human:8.00G
maxmemory_policy:noeviction
allocator_frag_ratio:-nan
allocator_frag_bytes:0
allocator_rss_ratio:-nan
allocator_rss_bytes:0
rss_overhead_ratio:-nan
rss_overhead_bytes:0
mem_fragmentation_ratio:-nan
mem_fragmentation_bytes:0
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_clients_slaves:0
mem_clients_normal:0
mem_aof_buffer:0
mem_allocator:jemalloc-5.1.0
active_defrag_running:0
lazyfree_pending_objects:0

# Persistence
loading:1
rdb_changes_since_last_save:0
rdb_bgsave_in_progress:0
rdb_last_save_time:1672391258
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0
module_fork_in_progress:0
module_fork_last_cow_size:0
loading_start_time:1672391258
loading_total_bytes:1089754870
loading_loaded_bytes:1088421887
loading_loaded_perc:99.88
loading_eta_seconds:0

# Stats
total_connections_received:6
total_commands_processed:6
instantaneous_ops_per_sec:0
total_net_input_bytes:352
total_net_output_bytes:246
instantaneous_input_kbps:0.00
instantaneous_output_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
expire_cycle_cpu_milliseconds:0
evicted_keys:0
keyspace_hits:0
keyspace_misses:0
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0
tracking_total_keys:0
tracking_total_items:0
tracking_total_prefixes:0
unexpected_error_replies:0
total_reads_processed:14
total_writes_processed:14
io_threaded_reads_processed:0
io_threaded_writes_processed:0

# Replication
role:master
connected_slaves:0

master_replid:2edb2ebbc3d0a755061aad0d2230139a464825cc
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:536870912
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:9.569255
used_cpu_user:85.149122
used_cpu_sys_children:0.000000
used_cpu_user_children:0.000000

# Modules
module:name=graph,ver=20806,api=1,filters=0,usedby=[],using=[],options=[]

# Commandstats
cmdstat_auth:calls=6,usec=40,usec_per_call=6.67

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=26,expires=0,avg_ttl=0
```

**Comment by LiorKogan** [comment]
You were using a beta (pre-GA) version of RedisGraph: 2.8.6
We would suggest that you upgrade to the latest patch (currently 2.8.21). Many issues have been solved since.
Also, upgrade your Redis version from 6.0.16 to 6.2.8

As for RDB recovery, unfortunately, there is no automated tool you can use.

**Comment by zhangliang-zl** [comment]
We have no plans to upgrade to the latest version anytime soon
Under what circumstances would such a thing arise? How should we avoid it?

**Comment by LiorKogan** [comment]
The root cause is most likely #2381 fixed in RedisGraph 2.8.14.
The exact conditions under which this can happen are hard to describe, but certain sequences of queries involving delete operations can lead to it.

Theoretically, you can scavenge the nodes/relationships that were not corrupted from this RDB, but you will have to implement your own tools/script for doing this.
